src/metabase/api/activity.clj

```diff
@@ -189,11 +190,77 @@
 
-#_{:clj-kondo/ignore [:deprecated-var]}
-(api/defendpoint-schema GET "/recent_views"
-  "Get the list of 5 things the current user has been viewing most recently."
+(defn- bookmarks
+  [user-id]
+  (let [as-null (when (= (mdb.connection/db-type) :postgres) (h2x/->integer nil))]
+    {:select    [[:type :model] [:item_id :model_id]]
+     :from [[{:union-all [{:select [:card_id
+                                    [as-null :dashboard_id]
+                                    [as-null :collection_id]
+                                    [:card_id :item_id]
+                                    [[:inline "card"] :type]
+                                    :created_at]
+                           :from   [:card_bookmark]
+                           :where  [:= :user_id [:inline user-id]]}
+                          {:select [[as-null :card_id]
+                                    :dashboard_id
+                                    [as-null :collection_id]
+                                    [:dashboard_id :item_id]
+                                    [[:inline "dashboard"] :type]
+                                    :created_at]
+                           :from   [:dashboard_bookmark]
+                           :where  [:= :user_id [:inline user-id]]}]}
+             :bookmarks]]}))
+
+(defn- recent-views-for-user
+  [user-id]
+  (let [bookmarks (bookmarks user-id)
+        qe        {:select [[[:inline "qe"] :source]
+                            [:executor_id :user_id]
+                            :context
+                            [:started_at :timestamp]
+                            [[:inline "card"] :model]
+                            [:card_id :model_id]
+                            [false :dataset]]
+                   :from   :query_execution}
+        vl        {:select    [[[:inline "vl"] :source]
+                               :user_id
+                               [[:inline "question"] :context]
+                               :timestamp
+                               :model
+                               :model_id
+                               [:report_card.dataset :dataset]]
+                   :from      [:view_log]
+                   :left-join [:report_card
+                               [:and
+                                [:= :view_log.model [:inline "card"]]
+                                [:= :view_log.model_id :report_card.id]]]}
+        views     {:union-all [qe vl]}]
+    (t2/query
+     {:select   [[[:max :timestamp] :timestamp]
+                 :model
+                 :model_id]
+      :from     [[views :views]]
+      :where    [[:and
+                  [:= :user_id [:inline user-id]]
+                  [:>= :timestamp
+                   (case (mdb.connection/db-type)
+                     :postgres [:- [:now] [:cast "30 days" :interval]]
+                     :h2 [:dateadd [:inline "DAY"] [-30] [:current_date]]
+                     [:>= :timestamp [:- [:date [:now]] [:inline 30]]])]
```

Comment:
I _think_ if you generate a cutoff timestamp in Clojure you can do a straight comparison against the `timestamp` row and it'll be the same for all DBs. Something like `(t/minus (t/offset-date-time) (t/days 30))`

Reply:
Oh my goodness, thank you! That's really good to know, and so much simpler